controller.go

```diff
@@ -64,40 +114,88 @@ func ServerLoop(listenPort int) {
 	port = listenPort
 
 	tcpAddr, err := net.ResolveTCPAddr("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		fmt.Printf("resolve tcp addr failed: %v", err)
+		return
+	}
 	listener, err := net.ListenTCP("tcp", tcpAddr)
-
-	ofc := NewOFController()
-	GetAppManager().RegistApplication(ofc)
-
 	if err != nil {
+		fmt.Printf("failed to listen: %v", err)
 		return
 	}
+	ofc := NewOFController()
+	ofc.RegisterHandler(h)
+	if err := ofc.Serve(listener); err != nil {
+		fmt.Printf("failed to serve: %v", err)
+		return
 
-	// wait for connect from switch
-	for {
-		conn, err := listener.AcceptTCP()
-		if err != nil {
-			return
-		}
-		go handleConnection(conn)
 	}
 }
 
+func (s *OFController) Close() error {
+	atomic.StoreInt32(&s.inShutdown, 1)
+	s.mtx.Lock()
+	err := s.listener.Close()
+	for _, dp := range s.dps {
+		dp.Close()
+	}
+	s.mtx.Unlock()
+	return err
+}
+
+func (s *OFController) Shutdown(ctx context.Context) error {
+	atomic.StoreInt32(&s.inShutdown, 1)
```

Comment:
do we want to be able to call Close() or Shutdown() multiple times? Shouldn't be there something like:
```suggestion
if s.isShuttingDown() {
return nil
}
	atomic.StoreInt32(&s.inShutdown, 1)
```

Reply:
I made that interface similar to http.Server. Functions Close and shutdown from http.Server doesn't have that checks.